packages/client/src/clients/in-app-messages/socket-driver.ts

```diff
@@ -0,0 +1,187 @@
+import { Store } from "@tanstack/store";
+import { Channel, Socket } from "phoenix";
+
+import { InAppMessagesClient } from "./message-client";
+import { InAppMessage, InAppMessagesClientOptions } from "./types";
+
+export enum SocketEventType {
+  MessageCreated = "message.created",
+}
+
+const SOCKET_EVENT_TYPES = [SocketEventType.MessageCreated];
+
+type ClientQueryParams = InAppMessagesClientOptions;
+
+// e.g. in_app:<message-type>:<channel_id>:<user_id>
+type ChannelTopic = string;
+
+// Unique reference id of an in-app message client
+type ClientReferenceId = string;
+
+type MessageCreatedEventPayload = {
+  event: SocketEventType.MessageCreated;
+  topic: string;
+  data: {
+    in_app_message: InAppMessage;
+  };
+};
+
+export type SocketEventPayload = MessageCreatedEventPayload;
+
+// "attn" field contains a list of client reference ids that should be notified
+// of a socket event.
+type WithAttn<P> = P & { attn: Array<ClientReferenceId> };
+
+type InAppMessageSocketInbox = Record<ClientReferenceId, SocketEventPayload>;
+
+/*
+ * Manages socket subscriptions for in-app messages, allowing multiple in-app
+ * message clients or components to listen for real time updates from the socket
+ * API via a single socket connection. It's expected to be instantiated once
+ * per an in-app channel.
+ */
+export class InAppMessageSocketDriver {
+  // Mapping of live channels by topic. Note, there can be one or more in-app
+  // message client(s) that can subscribe.
+  private channels: Record<ChannelTopic, Channel>;
+
+  // Mapping of query params for each in-app message client, partitioned by its
+  // reference id, and grouped by channel topic. It's a double nested object
+  // that looks like:
+  //  {
+  //    "in_app:card:...": {
```

Comment:
The inner object is what gets sent to the socket API as params: 
https://github.com/knocklabs/switchboard/pull/2963/files#diff-aa5f2245deb4ebc08de405e1f9cac816d0e942bcfffbfcd1693e8d3594d24427R30